pallets/crowdloans/src/lib.rs

```diff
@@ -531,6 +554,78 @@ pub mod pallet {
+            // AccountId, cTokenId, percentOfIssuance
+            let ownership_snapshot = Self::ctoken_ownership_snapshot();
+
+            // now for each account lets payout
+            for (account_id, accounts_ownership) in ownership_snapshot.iter() {
+                // payout for each vault
+                for (ctoken, percent_of_ctokens) in accounts_ownership.iter() {
+                    // get the total number of ptokens to be distributed
+                    let ptoken = Self::get_ptoken_from_ctoken(ctoken);
+                    let current_distribution_total = Self::get_distribution_amount(ptoken);
+
+                    // the user should recieve this portion of ptokens
+                    let amount = current_distribution_total * percent_of_ctokens;
+
+                    // transfer tokens to user from this account
+                    T::Assets::transfer(ptoken, &Self::account_id(), &account_id, amount, true)
+                        .map_err(|_: DispatchError| Error::<T>::InsufficientBalance)?;
+
+                    // emit event that we distributed
+                    Self::deposit_event(Event::<T>::DistributedProjectToken(
+                        ptoken,
+                        account_id.clone(),
+                        amount,
+                    ));
+                }
+            }
+            Ok(())
+        }
+
+        // TODO: how do we determine the ptoken for each ctoken
+        fn get_ptoken_from_ctoken(_ctoken: &u32) -> u32 {
+            0
+        }
+
+        // TODO: how do we determine the balance of tokens to distribute this iteration
+        fn get_distribution_amount(_ptoken: u32) -> u128 {
+            10_000
+        }
+
+        pub fn ctoken_ownership_snapshot() -> BTreeMap<T::AccountId, BTreeMap<u32, u128>> {
+            // iterate over all accounts on chain and build ownership map for each
+            frame_system::Account::<T>::iter()
+                .map(|(who, _account_information)| {
+                    // iterate over each vault and build a map of ctoken and percent
+                    let ctoken_ownership_map: BTreeMap<u32, u128> = Vaults::<T>::iter()
+                        .map(|(_para_id, vault)| {
+                            // get total issuance
+                            let ctoken_issuance = T::Assets::total_issuance(vault.ctoken);
+
+                            // get individuals balance
+                            let balance = T::Assets::balance(vault.ctoken, &who.clone());
+
+                            // percent of total ctokens
+                            let percent_of_payout = balance / ctoken_issuance;
+
+                            // vault.ctoken, who, percent_of_payout
+                            (vault.ctoken, percent_of_payout)
+                        })
+                        .collect(); // we collect in to a variable for readability
+                    (who, ctoken_ownership_map)
+                })
+                .collect()
+        }
+
```

Comment:
I am wondering if there could be a race condition here where if we take more than one block, or 12 seconds, to execute this function we may have a case where some of the balances change while this function is still running